crates/db_schema/src/impls/actor_language.rs

```diff
@@ -0,0 +1,350 @@
+      // Clear the current user languages
+      delete(local_user_language.filter(local_user_id.eq(for_local_user_id))).execute(conn)?;
+
+      let lang_ids = update_languages(conn, language_ids)?;
+      for l in lang_ids {
+        let form = LocalUserLanguageForm {
+          local_user_id: for_local_user_id,
+          language_id: l,
+        };
+        insert_into(local_user_language)
+          .values(form)
+          .get_result::<Self>(conn)?;
+      }
+      Ok(())
+    })
+  }
+}
+
+impl SiteLanguage {
+  pub fn read_local(conn: &PgConnection) -> Result<Vec<LanguageId>, Error> {
+    conn.build_transaction().read_write().run(|| {
+      let local_site = Site::read_local(conn)?;
+      SiteLanguage::read(conn, local_site.id)
+    })
+  }
+
+  pub fn read(conn: &PgConnection, for_site_id: SiteId) -> Result<Vec<LanguageId>, Error> {
+    use crate::schema::site_language::dsl::*;
+    site_language
+      .filter(site_id.eq(for_site_id))
+      .select(language_id)
+      .load(conn)
+  }
+
+  pub fn update(
+    conn: &PgConnection,
+    language_ids: Vec<LanguageId>,
+    for_site_id: SiteId,
+  ) -> Result<(), Error> {
+    conn.build_transaction().read_write().run(|| {
+      use crate::schema::site_language::dsl::*;
+      // Clear the current languages
+      delete(site_language.filter(site_id.eq(for_site_id))).execute(conn)?;
+
+      let lang_ids = update_languages(conn, language_ids)?;
+      for l in lang_ids.clone() {
+        let form = SiteLanguageForm {
+          site_id: for_site_id,
+          language_id: l,
+        };
+        insert_into(site_language)
+          .values(form)
+          .get_result::<Self>(conn)?;
+      }
+
+      CommunityLanguage::limit_languages(conn, lang_ids)?;
+
+      Ok(())
+    })
+  }
+}
+
+impl CommunityLanguage {
+  /// Returns true if the given language is one of configured languages for given community
+  pub fn is_allowed_community_language(
+    conn: &PgConnection,
+    for_language_id: LanguageId,
+    for_community_id: CommunityId,
+  ) -> Result<(), LemmyError> {
+    use crate::schema::community_language::dsl::*;
+    let is_allowed = select(exists(
+      community_language
+        .filter(language_id.eq(for_language_id))
+        .filter(community_id.eq(for_community_id)),
+    ))
+    .get_result(conn)?;
+
+    if is_allowed {
+      Ok(())
+    } else {
+      Err(LemmyError::from_message("language_not_allowed"))
+    }
+  }
+
+  /// When site languages are updated, delete all languages of local communities which are not
+  /// also part of site languages. This is because post/comment language is only checked against
+  /// community language, and it shouldnt be possible to post content in languages which are not
+  /// allowed by local site.
+  fn limit_languages(conn: &PgConnection, site_language_ids: Vec<LanguageId>) -> Result<(), Error> {
+    // this could be implemented using join + delete, but its not supported in diesel yet
```

Comment:
Diesel might not let you do join + delete, but they can let you do a left outer join and then loop over the results. See below.